### DISCUSSION TITLE
How to set check constraint in sql model?

### DISCUSSION THREAD

**Discussion by christianholland** [fastapi/sqlmodel#1951] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
pass
```

### Description

`SQLAlchemy` allows to implement [`CheckContraint`](https://docs.sqlalchemy.org/en/14/core/constraints.html?msclkid=ce2b48ddb19811ec80f36dc6060ceae0#check-constraint) on either column or table level. How can I archive the same in  `SQLModel`, e.g. limiting the possible values of a column to the colors `red` and `green` or to ensure that the values within a column are greater than `10`.

I suppose it must be somehow possible using `sa_column_kwargs` (similar to implementing `UniqueConstraint`).

Help would be highly appreciated, thanks!

### Operating System

Windows

### SQLModel Version

0.0.6

### Python Version

3.8.1

**Comment by jd-solanki** [comment]
You can use various sa_column* args like:
- `Field(sa_column_args=[CheckConstraint('col1>5')])`
-  `Field(sa_column=Column(...))`

**Comment by jd-solanki** [comment]
I use `sa_column` for Enum like below:
```python
    access_status: EnumAccessStatus = Field(
        sa_column=Column(
            SaEnumAccessStatus,
            nullable=False,
        )
    )
```

**Comment by dannyrohde** [comment]
One caveat to using columns is that you might need to move properties like `index` or `nullable` into the Column description. I found that it otherwise doesn't get honoured.

E.g. the example in the above comment will work, but the below **won't have the intended effect**:

`access_status: EnumAccessStatus = Field( sa_column=Column(SaEnumAccessStatus,), nullable=False, index=True)`

**Comment by LumaC0** [comment]
How about `CheckConstraint` at the table level? I've been setting it in __table_args__ like I'd do in sqlalchemy and just noticed Alembic isn't recognizing it.

**Comment by YuriiMotov** [comment]
> One caveat to using columns is that you might need to move properties like `index` or `nullable` into the Column description. I found that it otherwise doesn't get honoured.
> E.g. the example in the above comment will work, but the below **won't have the intended effect**:
> `access_status: EnumAccessStatus = Field( sa_column=Column(SaEnumAccessStatus,), nullable=False, index=True)`

In current version, you will get a runtime error if you try using `nullable` or `index` with `sa_column`:
```
RuntimeError: Passing nullable is not supported when also passing a sa_column
```

Considering [this comment from Sebastian](https://github.com/fastapi/sqlmodel/pull/436#issuecomment-1784034309)

> So I have intentionally simplified and limited the things I support in SQLModel, to work for the majority of use cases (say 90%) in a very simple and intuitive way (at least I hope so) without all the more advanced/exotic features that SQLAlchemy supports.

I think that `sa_column` is the only way to go..

**Comment by smiling-watermelon** [comment]
I know that that's pretty tangential to the question at hand, but, has anyone else encountered a problem where by setting `sa_column` after you've had a field defined without `sa_column`, when making a migration with alembic, sqlalchemy will raise an exception saying something like:

`sqlalchemy.exc.ArgumentError: Column object 'amount' already assigned to Table 'balance'`

I've encountered this error while setting CheckConstraint, so, it's sort of related to the question posted here, but not really 😅 


UPD: This is related to the inheritance. I suppose table is set on Column and consequently - Field on definition, so, if you have a base model (like I had for balance), you'll have to duplicate the field in your child models 😢

**Comment by ltianyi992** [comment]
You can set `CheckConstraint` in SQLModel by passing it through `__table_args__` for table-level constraints, or via `sa_column` for column-level constraints.

**Table-level CheckConstraint (most common):**

```python
from sqlalchemy import CheckConstraint
from sqlmodel import SQLModel, Field

class Item(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    color: str = Field(max_length=10)
    quantity: int

    __table_args__ = (
        CheckConstraint("color IN ('red', 'green')", name="ck_item_color"),
        CheckConstraint("quantity > 10", name="ck_item_quantity"),
    )
```

**Column-level CheckConstraint via `sa_column`:**

```python
from sqlalchemy import Column, Integer, CheckConstraint
from sqlmodel import SQLModel, Field

class Item(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    quantity: int = Field(
        sa_column=Column(
            Integer,
            CheckConstraint("quantity > 10", name="ck_item_quantity"),
        )
    )
```

**For enum-style constraints**, consider using a Python `Enum` type instead of a string column — SQLModel/SQLAlchemy will generate a proper DB-level enum (or a CHECK constraint depending on the backend):

```python
from enum import Enum
from sqlmodel import SQLModel, Field

class Color(str, Enum):
    red = "red"
    green = "green"

class Item(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    color: Color = Field(default=Color.red)
```

The `__table_args__` approach (first example) is the most flexible and works for any constraint expression, including cross-column checks like `CHECK(end_date > start_date)`.
